openshift-kube-apiserver/filters/deprecatedapirequest/controller.go

```diff
@@ -0,0 +1,239 @@
+}
+
+func (c *controller) ensureLogExists(ctx context.Context, r string) error {
+	_, err := c.client.Get(ctx, r, metav1.GetOptions{})
+	if err == nil {
+		return nil
+	}
+	if !errors.IsNotFound(err) {
+		return err
+	}
+	log := &apiv1.DeprecatedAPIRequest{
+		ObjectMeta: metav1.ObjectMeta{Name: r},
+		Spec:       apiv1.DeprecatedAPIRequestSpec{RemovedRelease: c.removedRelease(r)},
+	}
+	_, err = c.client.Create(ctx, log, metav1.CreateOptions{})
+	return err
+}
+
+type requestCounts struct {
+	sync.Mutex
+	LastUpdate time.Time
+	Counts     map[string]map[int]map[string]map[string]int
+}
+
+func (rc *requestCounts) IncrementRequestCount(r, verb, user string, ts time.Time, count int) {
+	rc.incrementRequestCount(r, verb, user, ts.Hour(), count)
+	if ts.After(rc.LastUpdate) {
+		rc.LastUpdate = ts
+	}
+}
+
+func (rc *requestCounts) incrementRequestCount(r, verb, user string, h, count int) {
+	if rc.Counts == nil {
+		rc.Counts = map[string]map[int]map[string]map[string]int{}
+	}
+	if _, ok := rc.Counts[r]; !ok {
+		rc.Counts[r] = map[int]map[string]map[string]int{}
+	}
+	if _, ok := rc.Counts[r][h]; !ok {
+		rc.Counts[r][h] = map[string]map[string]int{}
+	}
+	if _, ok := rc.Counts[r][h][user]; !ok {
+		rc.Counts[r][h][user] = map[string]int{}
+	}
+	if _, ok := rc.Counts[r][h][user][verb]; !ok {
+		rc.Counts[r][h][user][verb] = 0
+	}
+	rc.Counts[r][h][user][verb] += count
+}
+
+func (rc *requestCounts) IncrementRequestCounts(o *requestCounts) {
+	for r := range o.Counts {
+		for h := range o.Counts[r] {
+			for u := range o.Counts[r][h] {
+				for v := range o.Counts[r][h][u] {
+					rc.incrementRequestCount(r, v, u, h, o.Counts[r][h][u][v])
+				}
+			}
+		}
+	}
+	if o.LastUpdate.After(rc.LastUpdate) {
+		rc.LastUpdate = o.LastUpdate
+	}
+}
+
+func (rc *requestCounts) trim(now time.Time) {
+	// clear all if no updates in the last 24 hours
+	if rc.LastUpdate.Before(now.Add(-24 * time.Hour)) {
+		rc.Counts = map[string]map[int]map[string]map[string]int{}
+		rc.LastUpdate = now
+		return
+	}
+	// clear updates from about 24 hours ago
+	h := (now.Hour() + 1) % 24
+	for r := range rc.Counts {
+		if u, ok := rc.Counts[r][h]; ok && len(u) > 0 {
+			delete(rc.Counts[r], h)
+		}
+	}
+	// limit number of users to ten, in order to keep memory usage bound
+	for r := range rc.Counts {
+		for h := range rc.Counts[r] {
+			if len(rc.Counts[r][h]) > 10 {
+				// make list of users
+				users := make([]string, 0, len(rc.Counts[r][h]))
+				for u := range rc.Counts[r][h] {
+					users = append(users, u)
+				}
+				// sort by total request counts (descending)
+				sort.Slice(users, func(i, j int) bool {
+					ic, jc := 0, 0
+					for v := range rc.Counts[r][h][users[i]] {
+						ic += rc.Counts[r][h][users[i]][v]
+					}
+					for v := range rc.Counts[r][h][users[j]] {
+						jc += rc.Counts[r][h][users[j]][v]
+					}
+					return ic > jc
+				})
+				// keep only top 10
+				users = users[10:]
+				for _, u := range users {
+					delete(rc.Counts[r][h], u)
```

Comment:
if this method is called every 5 minutes let's say.  Then we would not be tracking the counts for maximum usage in an hour.  A user making 1 request a minute will make 60 per hour, but could get drowned out by a bursty actor every five minutes.  Let's make sure we keep the property of most usage in an hour.